src/screen/dashboard.rs

```diff
@@ -1659,60 +1617,90 @@ impl Dashboard {
     }
 
-    fn get_focused_mut(
-        &mut self,
-        main_window: &Window,
-    ) -> Option<(window::Id, pane_grid::Pane, &mut Pane)> {
+    fn get_focused_mut(&mut self) -> Option<(window::Id, pane_grid::Pane, &mut Pane)> {
         let (window, pane) = self.focus?;
         self.panes
-            .get_mut(main_window.id, window, pane)
+            .get_mut(window, pane)
             .map(|state| (window, pane, state))
     }
 
     fn get_focused_with_history_mut(
         &mut self,
-        main_window: &Window,
     ) -> Option<(pane_grid::Pane, &mut Pane, &mut history::Manager)> {
         let (window, pane) = self.focus?;
         self.panes
-            .get_mut(main_window.id, window, pane)
+            .get_mut(window, pane)
             .map(|state| (pane, state, &mut self.history))
     }
 
     pub fn get_unique_queries(&self, server: &Server) -> Vec<&target::Query> {
         self.history.get_unique_queries(server)
     }
 
-    fn focus_pane(
-        &mut self,
-        main_window: &Window,
-        window: window::Id,
-        pane: pane_grid::Pane,
-    ) -> Task<Message> {
-        if self.focus != Some((window, pane)) {
-            self.focus = Some((window, pane));
-
-            if let Some(task) = self.panes.iter(main_window.id).find_map(|(w, p, state)| {
+    pub fn focus_focused_pane_buffer(&mut self) -> Task<Message> {
+        if let Some((window, pane)) = self.focus {
+            if let Some(focus_buffer) = self.panes.iter().find_map(|(w, p, state)| {
                 (w == window && p == pane).then(|| {
                     state.buffer.focus().map(move |message| {
                         Message::Pane(window, pane::Message::Buffer(pane, message))
                     })
                 })
             }) {
-                return Task::batch(vec![task, window::gain_focus(window)]);
+                return focus_buffer;
             }
         }
 
         Task::none()
     }
 
+    fn focus_pane(&mut self, window: window::Id, pane: pane_grid::Pane) -> Task<Message> {
+        let task = if self.focus != Some((window, pane)) {
+            self.focus = Some((window, pane));
+
+            self.last_changed = Some(Instant::now());
+
+            if window == self.main_window() {
+                self.focus_history.push_front(pane);
+
+                self.focus_history.truncate(FOCUS_HISTORY_LEN);
+            }
+
+            window::gain_focus(window)
+        } else {
+            Task::none()
+        };
+
+        task.chain(self.focus_focused_pane_buffer())
+    }
+
+    pub fn focus_first_pane(&mut self, window: window::Id) -> Task<Message> {
+        let pane = self
+            .panes
+            .iter()
+            .find_map(|(w, pane, _)| (w == window).then_some(pane));
+
+        pane.map_or(Task::none(), |pane| self.focus_pane(window, pane))
+    }
+
+    pub fn focus_last_focused_or_first_pane(&mut self, window: window::Id) -> Task<Message> {
```

Comment:
Nit: Maybe `focus_window`? Since we seem to call this only when window focus changes.